dora/core/client/fs/src/test/java/alluxio/client/file/dora/WorkerLocationPolicyTest.java

```diff
@@ -0,0 +1,196 @@
+
+import com.google.common.collect.ImmutableList;
+import org.apache.commons.lang3.RandomStringUtils;
+import org.junit.Test;
+
+import java.util.List;
+import java.util.NavigableMap;
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class WorkerLocationPolicyTest {
+  private static final long WORKER_LIST_TTL_MS = 20;
+  private static final String OBJECT_KEY = "/path/to/object";
+  private static final int NUM_VIRTUAL_NODES = 100;
+
+  @Test
+  public void uninitializedThrowsException() {
+    ConsistentHashProvider provider = new ConsistentHashProvider(1, WORKER_LIST_TTL_MS);
+    assertThrows(IllegalStateException.class, () -> provider.get(OBJECT_KEY, 0));
+  }
+
+  @Test
+  public void concurrentInitialization() {
+    ConsistentHashProvider provider = new ConsistentHashProvider(1, WORKER_LIST_TTL_MS);
+    final int numThreads = 16;
+    CountDownLatch startSignal = new CountDownLatch(numThreads);
+    ExecutorService executorService = Executors.newFixedThreadPool(numThreads);
+    List<List<BlockWorkerInfo>> lists = IntStream.range(0, numThreads)
+        .mapToObj(i -> generateRandomWorkerList(5))
+        .collect(Collectors.toList());
+    List<Future<NavigableMap<Integer, BlockWorkerInfo>>> futures = IntStream.range(0, numThreads)
+        .mapToObj(i -> {
+          List<BlockWorkerInfo> list = lists.get(i);
+          return executorService.submit(() -> {
+            startSignal.countDown();
+            try {
+              startSignal.await();
+            } catch (InterruptedException e) {
+              fail("interrupted");
+            }
+            provider.refresh(list, NUM_VIRTUAL_NODES);
+            return provider.getActiveNodesMap();
+          });
+        })
+        .collect(Collectors.toList());
+    Set<NavigableMap<Integer, BlockWorkerInfo>> mapSet = futures.stream().map(future -> {
+      try {
+        return future.get();
+      } catch (InterruptedException interruptedException) {
+        throw new AssertionError("interrupted", interruptedException);
+      } catch (ExecutionException e) {
+        throw new AssertionError("failed to run thread", e);
+      }
+    }).collect(Collectors.toSet());
+
+    assertEquals(1, mapSet.size());
+    // check if the worker list is one of the lists provided by the threads
+    List<BlockWorkerInfo> workerInfoListUsedByPolicy = provider.getLastWorkerInfos();
+    assertTrue(lists.contains(workerInfoListUsedByPolicy));
+    assertEquals(
+        ConsistentHashProvider.build(workerInfoListUsedByPolicy, NUM_VIRTUAL_NODES),
+        provider.getActiveNodesMap());
+  }
+
+  @Test
+  // todo(bowen): this test can be flaky if the test subject is not thread safe
```

Comment:
if this is sure to be flaky, we shall not include any known flaky test

Reply:
the tricky thing about this test is that it's trying to ensure the test subject is thread safe. When it is, the test is not flaky. When it is not, then it can be flaky. I can't think of a way to reliably detect if the test is thread safe or not.

Reply:
added comments on the test flakiness